### PR TITLE
feat: add MTV operator upgrade/update migration testing

### PR DESCRIPTION
**Pull request by krcmarik**
Add upgrade test infrastructure:
- test_upgrade_migration.py: incremental test class that creates migration resources, upgrades MTV operator, verifies post-upgrade state, and runs cold migration
- utilities/upgrade.py: subprocess-based upgrade runner with fd 3 support for mtv-autodeploy logging, plus pod readiness and plan verification
- Add @pytest.mark.upgrade to warm migration for post-upgrade warm test
- Remove tier0 from cold migration (covered by upgrade test)
- Add upgrade_script_path, mtv_upgrade_to_version, mtv_upgrade_to_source, mtv_upgrade_image_index config parameters
- Add 'upgrade' marker to pytest.ini
- Add MtvUpgradeError and ForkliftPodsNotRunningError exceptions

## Summary by CodeRabbit

* **New Features**
  * End-to-end upgrade testing that automates operator upgrade and cold-migration validation.

* **Tests**
  * Added a dedicated upgrade test suite and a new "upgrade" pytest marker; applied marker to existing warm-migration sanity tests.
  * New fixtures to prepare pre-upgrade resources and provide the upgrade script path.

* **Chores**
  * Configurable upgrade parameters, an upgrade runner utility, a specific upgrade error type, added git runtime support, and test docs/usage examples.

[Review Change Stack](https://app.coderabbit.ai/change-stack/RedHatQE/mtv-api-tests/pull/487?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: utilities/upgrade.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+from exceptions.exceptions import ForkliftPodsNotRunningError, MtvUpgradeError
+
+if TYPE_CHECKING:
+    from kubernetes.dynamic import DynamicClient
+
+LOGGER = get_logger(name=__name__)
+
+
+@contextmanager
+def _duplicate_stdout_to_fd3() -> Generator[None]:
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply utilities/upgrade.py]: Done

File: tests/upgrade/test_upgrade_migration.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            target_namespace=target_namespace,
+        )
+
+    def test_check_vms(
+        self,
+        prepared_plan: dict[str, Any],
+        source_provider: BaseProvider,
+        destination_provider: BaseProvider,
+        source_provider_data: dict[str, Any],
+        target_namespace: str,
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/upgrade/test_upgrade_migration.py]: Done

**Comment by krcmarik** [comment]
@coderabbitai review

File: utilities/upgrade.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                env=env,
+                cwd=os.path.dirname(script_path),
+                check=True,
+                pass_fds=(3,),
+            )
+    except subprocess.CalledProcessError as exc:
+        raise MtvUpgradeError(f"MTV upgrade script failed with exit code {exc.returncode}") from exc
+
+    LOGGER.info(f"MTV upgrade to version {mtv_version} completed successfully")
+
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply utilities/upgrade.py]: Done

**Comment by krcmarik** [comment]
@coderabbitai review

File: tests/upgrade/test_upgrade_migration.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+)
+@pytest.mark.usefixtures("cleanup_migrated_vms")
+class TestUpgradeColdMigration:
+    """Cold migration with MTV operator upgrade between plan creation and execution."""
+
+    storage_map: StorageMap
+    network_map: NetworkMap
+    plan_resource: Plan
+
+    def test_verify_pre_upgrade_version(
```
Comment (myakove) [inline tests/upgrade/test_upgrade_migration.py]: this should be part of the setup, this is prerequsite and not a test

File: tests/upgrade/test_upgrade_migration.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            expected = Version(expected_version)
+        except InvalidVersion as exc:
+            raise AssertionError(
+                f"Could not parse MTV version strings: actual='{version}', expected='{expected_version}'"
+            ) from exc
+        assert (actual.major, actual.minor) == (expected.major, expected.minor), (
+            f"Expected MTV major.minor version '{expected.major}.{expected.minor}', got '{version}'"
+        )
+
+        wait_for_forklift_pods_ready(admin_client=ocp_admin_client)
```
Comment (myakove) [inline tests/upgrade/test_upgrade_migration.py]: the upgrade script verify all pods are running,why this is needed here as well?

File: utilities/upgrade.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    mtv_version: str,
+    mtv_source: str,
+    image_index: str = "",
+) -> None:
+    """Run the MTV operator upgrade using the specified upgrade script.
+
+    Cluster credentials are read from py_config and passed as environment variables.
+
+    Args:
+        script_path (str): Full path to the upgrade script (e.g., "/path/to/mtv-autodeploy/mtv-upgrade.sh").
```
Comment (myakove) [inline utilities/upgrade.py]: why do we have this? as I see it we should have a ffixture that clone the repo to pytest tmppath and carry on to this function as arg to use

File: utilities/upgrade.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        raise ValueError("mtv_source must be provided via --tc=mtv_upgrade_to_source:<source>")
+
+    env = os.environ.copy()
+    env.update({
+        "MTV_VERSION": mtv_version,
+        "MTV_SOURCE": mtv_source.upper(),
+        "IMAGE_INDEX": image_index,
+        "CLUSTER_USERNAME": py_config["cluster_username"],
+        "CLUSTER_PASSWORD": py_config["cluster_password"],
+        "CLUSTER_API_URL": py_config["cluster_host"],
```
Comment (myakove) [inline utilities/upgrade.py:44]: we already have client object that contain all this info and we should use it

File: utilities/upgrade.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+from exceptions.exceptions import ForkliftPodsNotRunningError, MtvUpgradeError
+
+if TYPE_CHECKING:
+    from kubernetes.dynamic import DynamicClient
+
+LOGGER = get_logger(name=__name__)
+
+
+@contextmanager
+def _duplicate_stdout_to_fd3() -> Generator[None, None, None]:
```
Comment (myakove) [inline utilities/upgrade.py]: I think it better to make mtv upgrade script output logs to file (optional) then use it

File: utilities/upgrade.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        _admin_client=admin_client,
+        sleep=5,
+        wait_timeout=timeout,
+        exceptions_dict={ForkliftPodsNotRunningError: [], NotFoundError: []},
+    ):
+        if sample:
+            return
+
+
+def verify_plan_ready_after_upgrade(
```
Comment (myakove) [inline utilities/upgrade.py]: we already have code to check if plan is ready right? 
use it, there is nothing spaciel about waiting for plan before or after upgrade or without upgrade at all.

**Review by myakove — CHANGES_REQUESTED** [review]

File: utilities/upgrade.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+    try:
+        subprocess.run(
+            [script_path],
+            env=env,
+            cwd=os.path.dirname(script_path),
+            check=True,
+        )
+    except subprocess.CalledProcessError as exc:
+        raise MtvUpgradeError(f"MTV upgrade script failed with exit code {exc.returncode}") from exc
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply utilities/upgrade.py]: Done

**Comment by krcmarik** [comment]
@coderabbitai review

File: utilities/upgrade.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+            [script_path],
+            env=env,
+            cwd=os.path.dirname(script_path),
+            check=True,
+            timeout=3600,
+        )
+    except subprocess.TimeoutExpired as exc:
+        raise MtvUpgradeError(f"MTV upgrade script timed out after {exc.timeout} seconds") from exc
+    except subprocess.CalledProcessError as exc:
+        raise MtvUpgradeError(f"MTV upgrade script failed with exit code {exc.returncode}") from exc
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply utilities/upgrade.py]: 1. OSError catch: The function already guards against the common OSError cases — if not script_path catches empty paths, if not os.path.isfile(script_path) catches missing files (preventing FileNotFoundError).
   A PermissionError (file exists but not executable) is a real infrastructure/setup bug that should crash loudly — wrapping it in MtvUpgradeError would hide the actual problem. The docstring documents intended
  error cases, not every possible exception Python can throw.
  2. os.path.dirname("") returning "": Already impossible because if not script_path rejects empty strings and if not os.path.isfile(script_path) rejects bare filenames that don't resolve to a file. The path
  comes from py_config["upgrade_script_path"] (CI-controlled, always a full path)

File: utilities/upgrade.py (hunk 7 above)
Comment (krcmarik) [reply utilities/upgrade.py:44]: Just need to export the values into env for the upgrade script, afaik It's nt possible to extract the username and password from the Dynamic client object which holds the ocp token.

File: utilities/upgrade.py (hunk 6 above)
Comment (krcmarik) [reply utilities/upgrade.py]: I did not want to let the test suite to handle (clone) a different repo and rather leave it to pytest executor (.i.e. jenkins) and then just pass directly the upgrade script path but If you think It's better to do that in a fixture we can do that

File: utilities/upgrade.py (hunk 7 above)
Comment (myakove) [reply utilities/upgrade.py:44]: ack

**Comment by redhat-qe-bot1** [comment]
**Clean rebase detected** — no code changes compared to previous head (`ae714f6`).

**Comment by krcmarik** [comment]
@coderabbitai review

File: tests/upgrade/conftest.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        raise ValueError("upgrade_script_path must be provided via --tc=upgrade_script_path:<relative-path>")
+
+    clone_dir = tmp_path_factory.mktemp("mtv-autodeploy")
+
+    LOGGER.info(f"Cloning {repo_url} (ref={repo_ref}) into {clone_dir}")
+    Repo.clone_from(url=repo_url, to_path=str(clone_dir), branch=repo_ref, depth=1)
+
+    script_path = clone_dir / script_relative_path
+    if not script_path.is_file():
+        raise ValueError(f"Upgrade script not found at '{script_path}' after cloning {repo_url}")
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/upgrade/conftest.py:63]: script_relative_path comes from py_config["upgrade_script_path"], which is set via --tc= in Jenkins or by a developer running tests locally. Both are trusted contexts. There's no untrusted input boundary here
  — anyone who can set this config value already has full shell access to the machine. Path traversal prevention is for web requests and user uploads, not internal test configuration.

  The existing if not script_path.is_file() check already catches any misconfigured path (absolute, traversal, typo — all fail the same way with a clear ValueError). Adding security theatre validation would
  violate the project's "trust required arguments" and "don't add validation for scenarios that can't happen" principles.

File: tests/tests_config/config.py [hunk 13]
```diff
@@ -13,6 +13,12 @@
 snapshots_interval: int = 2
 mins_before_cutover: int = 5
 plan_wait_timeout: int = 3600
+upgrade_repo_url: str = ""
+upgrade_repo_ref: str = ""
+upgrade_script_path: str = ""
```
Comment (myakove) [inline tests/tests_config/config.py]: I still do not understand why we need this?

File: tests/upgrade/test_upgrade_migration.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+            destination_provider=destination_provider,
+            storage_map=self.storage_map,
+            network_map=self.network_map,
+            virtual_machines_list=prepared_plan["virtual_machines"],
+            target_namespace=target_namespace,
+            warm_migration=prepared_plan.get("warm_migration", False),
+        )
+        assert self.plan_resource, "Plan creation failed"
+
+    def test_upgrade_mtv(self, upgrade_script_path: str) -> None:
```
Comment (myakove) [inline tests/upgrade/test_upgrade_migration.py]: all the above should be setup, we do not test here the create of things on the current version.
the tests are the upgrade andall whats come after

**Review by myakove — CHANGES_REQUESTED** [review]

File: tests/tests_config/config.py (hunk 13 above)
Comment (krcmarik) [reply tests/tests_config/config.py]: The goal is to let an user to choose the upgrade script they want to use, should we use a hard-coded name (relative path within the repo) of the script?

File: tests/upgrade/test_upgrade_migration.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+        )
+    ],
+    indirect=True,
+    ids=["upgrade-cold"],
+)
+@pytest.mark.usefixtures("cleanup_migrated_vms")
+class TestUpgradeColdMigration:
+    """Cold migration with MTV operator upgrade between plan creation and execution."""
+
+    def test_upgrade_mtv(self, upgrade_script_path: str, upgrade_plan_resource: Plan) -> None:
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/upgrade/test_upgrade_migration.py]: @coderabbitai upgrade_plan_resource is used by value in test_verify_post_upgrade (calls .wait_for_condition()) and test_migrate_vms (passes it to execute_migration()). Keeping it as a parameter in test_upgrade_mtv too is consistent across the class — all three methods receive it the same way. Moving only one to usefixtures while the others keep it as a parameter would be inconsistent for no benefit.

File: tests/upgrade/conftest.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+    Returns:
+        str: Absolute path to the executable upgrade script.
+
+    Raises:
+        ValueError: If required config values are missing or the script is not found after cloning.
+        git.exc.GitCommandError: If the git clone fails.
+    """
+    repo_url: str = py_config.get("upgrade_repo_url", "")
+    repo_ref: str = py_config.get("upgrade_repo_ref", "")
+    script_relative_path: str = py_config.get("upgrade_script_path", "")
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/upgrade/conftest.py:47]: @coderabbitai The empty string isn't a silent fallback — it's a sentinel for the validation block that immediately follows and raises ValueError with the exact --tc= flag syntax needed. Bracket access would give a raw KeyError: 'upgrade_repo_url' which tells the user nothing about how to fix it. The .get() + if not pattern handles both missing keys and empty values with an actionable error message.

File: tests/tests_config/config.py (hunk 13 above)
Comment (myakove) [reply tests/tests_config/config.py]: I see the point, im fine with it.
next tiem please do not relove it till we get aggrement on the issue

File: tests/upgrade/conftest.py [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+        raise ValueError(f"Upgrade script not found at '{script_path}' after cloning {repo_url}")
+
+    current_mode = script_path.stat().st_mode
+    script_path.chmod(current_mode | stat.S_IXUSR)
+
+    return str(script_path)
+
+
+@pytest.fixture(scope="class")
+def upgrade_storage_map(
```
Comment (myakove) [inline tests/upgrade/conftest.py]: the name is confusing, we so not upgrade the storage map. `storage_map_for_upgrade` (or something else if you have a better one) is better I think. 
same for others

**Review by myakove — CHANGES_REQUESTED** [review]

File: tests/tests_config/config.py (hunk 13 above)
Comment (krcmarik) [reply tests/tests_config/config.py]: I thought that the problem was to introduce these parameters in config.py which is not necessary - these are not really test configuration parameters and are empty (never used from the config file itself). I would say It's more of operational/infrastructure parameters.

File: tests/upgrade/conftest.py (hunk 17 above)
Comment (krcmarik) [reply tests/upgrade/conftest.py]: Done. Renamed all three fixtures to `pre_upgrade_storage_map`, `pre_upgrade_network_map`, and `pre_upgrade_plan_resource` — emphasizes these resources are created before the upgrade.

File: tests/upgrade/conftest.py [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+    if not repo_ref:
+        raise ValueError("upgrade_repo_ref must be provided via --tc=upgrade_repo_ref:<ref>")
+    if not script_relative_path:
+        raise ValueError("upgrade_script_path must be provided via --tc=upgrade_script_path:<relative-path>")
+
+    clone_dir = tmp_path_factory.mktemp("mtv-autodeploy")
+
+    LOGGER.info(f"Cloning {repo_url} (ref={repo_ref}) into {clone_dir}")
+    Repo.clone_from(url=repo_url, to_path=str(clone_dir), branch=repo_ref, depth=1)
+
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/upgrade/conftest.py:60]: Done. Added upgrade test documentation to README.md with required `--tc` parameters, marker info, and example command.

File: tests/upgrade/conftest.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/upgrade/conftest.py:47]: The `.get("key", "")` pattern here is intentional — the empty string is not a silent default/fallback. It's a sentinel for the validation block that immediately follows (`if not repo_url: raise ValueError(...)`), which provides an actionable error message with the exact `--tc=` flag syntax needed. Bracket access would give a raw `KeyError: 'upgrade_repo_url'` which tells the user nothing about how to fix it. The `.get()` + `if not` pattern handles both missing keys and empty values with clear guidance.

File: tests/upgrade/test_upgrade_migration.py [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+        source_provider: BaseProvider,
+        destination_provider: BaseProvider,
+        source_provider_data: dict[str, Any],
+        source_vms_namespace: str,
+        source_provider_inventory: ForkliftInventory,
+        vm_ssh_connections: SSHConnectionManager | None,
+        pre_upgrade_network_map: NetworkMap,
+        pre_upgrade_storage_map: StorageMap,
+    ) -> None:
+        """Validate migrated VMs after upgrade."""
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/upgrade/test_upgrade_migration.py:118]: Test method parameters are pytest fixtures injected by the framework, not caller-provided arguments. Adding `Args:` sections listing fixtures would duplicate what's already visible in the function signature and would need to be updated whenever fixture signatures change. The existing one-line docstrings describing the test step are consistent with how test methods are documented throughout the codebase.

File: tests/upgrade/test_upgrade_migration.py [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+    """Cold migration with MTV operator upgrade between plan creation and execution."""
+
+    def test_upgrade_mtv(self, upgrade_script_path: str, pre_upgrade_plan_resource: Plan) -> None:
+        """Upgrade the MTV operator to the target version."""
+        run_mtv_upgrade(
+            script_path=upgrade_script_path,
+            mtv_version=py_config["mtv_upgrade_to_version"],
+            mtv_source=py_config["mtv_upgrade_to_source"],
+            image_index=py_config["mtv_upgrade_image_index"],
+        )
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/upgrade/test_upgrade_migration.py:61]: These keys are intentionally not defined in `config.py`. They are runtime-only parameters provided via `--tc=key:value` on the command line. The human reviewer specifically requested removing placeholder defaults from `config.py` for these keys. pytest-testconfig's `--tc` flag injects keys directly into `py_config` without requiring them to be pre-defined — this was verified. A `KeyError` on missing keys is the expected fail-fast behavior when upgrade tests are run without the required `--tc` flags.

File: utilities/upgrade.py [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+        stdout = exc.stdout.decode() if exc.stdout else ""
+        stderr = exc.stderr.decode() if exc.stderr else ""
+        raise MtvUpgradeError(
+            f"MTV upgrade script failed with exit code {exc.returncode}\nstdout: {stdout}\nstderr: {stderr}"
+        ) from exc
+
+    LOGGER.info(f"MTV upgrade stdout:\n{result.stdout.decode()}")
+    if result.stderr:
+        LOGGER.warning(f"MTV upgrade stderr:\n{result.stderr.decode()}")
+
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply utilities/upgrade.py:74]: Done. Added inline `.replace(env["CLUSTER_PASSWORD"], "***")` to all stdout/stderr logging and exception messages to redact credentials.

File: tests/upgrade/test_upgrade_migration.py [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+        assert (actual.major, actual.minor) == (expected.major, expected.minor), (
+            f"Expected MTV major.minor version '{expected.major}.{expected.minor}', got '{version}'"
+        )
+
+        pre_upgrade_plan_resource.wait_for_condition(
+            condition=Plan.Condition.READY,
+            status=Plan.Condition.Status.TRUE,
+            timeout=300,
+        )
+        LOGGER.info(f"Plan '{pre_upgrade_plan_resource.name}' is in Ready state after upgrade")
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/upgrade/test_upgrade_migration.py:89]: The upgrade script itself is responsible for ensuring pod readiness after the upgrade completes — it waits for all operator pods to be stable before returning. Adding a redundant pod check in the test would duplicate the script's responsibility. The test's role is to verify the MTV version and Plan CR state post-upgrade, not to re-validate infrastructure the upgrade script already guarantees.

File: tests/upgrade/test_upgrade_migration.py [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+            plan=prepared_plan,
+            source_provider=source_provider,
+            destination_provider=destination_provider,
+            network_map_resource=pre_upgrade_network_map,
+            storage_map_resource=pre_upgrade_storage_map,
+            source_provider_data=source_provider_data,
+            source_vms_namespace=source_vms_namespace,
+            source_provider_inventory=source_provider_inventory,
+            vm_ssh_connections=vm_ssh_connections,
+        )
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/upgrade/test_upgrade_migration.py:129]: The upgrade test intentionally deviates from the standard 5-step pattern. The key difference: migration resources (StorageMap, NetworkMap, Plan) must be created *before* the MTV operator upgrade to verify they survive the upgrade. The standard pattern creates resources inside test methods, but here they must exist pre-upgrade as class-scoped fixtures. The test flow is: create resources (fixtures) → upgrade operator → verify resources survived → execute migration → validate VMs. This is a fundamentally different workflow that doesn't fit the standard test structure.

File: tests/upgrade/conftest.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/upgrade/conftest.py:60]: Done. Added `git` to the Dockerfile runtime stage packages. GitPython shells out to the `git` CLI at runtime (it's a wrapper, not a pure Python implementation).

**Comment by krcmarik** [comment]
@coderabbitai review

File: tests/warm/test_mtv_warm_migration.py [hunk 24]
```diff
@@ -19,6 +19,7 @@
 @pytest.mark.rhv
 @pytest.mark.tier0
 @pytest.mark.warm
+@pytest.mark.upgrade
```
[the remark this answers is not in the document]
Comment (krcmarik) [reply tests/warm/test_mtv_warm_migration.py:22]: @qodo-code-review That's intentional trade off and migration should survive an operator restart anyway.

**Review by myakove — APPROVED** [review]
